README.md
<div align="center"><img src="/screens/gif_cover.gif"/></div>

# RotatingText
[![platform](https://img.shields.io/badge/Platform-Android-yellow.svg?style=flat-square)](https://www.android.com)
[![API](https://img.shields.io/badge/API-16%2B-brightgreen.svg?style=flat-square)](https://android-arsenal.com/api?level=16s)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square)](https://opensource.org/licenses/MIT)
[![By-MDG](https://img.shields.io/badge/By-MDG-orange.svg?style=flat-square)](http://mdg.iitr.ac.in)

Rotating text is an Android library that can be used to make text switching painless and beautiful, with the use of interpolators, typefaces and more customisations.

# Usage
Just add the following dependency in your app's `build.gradle`
```
dependencies {
      compile 'com.sdsmdg.harjot:rotatingtext:1.0.2'
}
```

## Example Usage 1 (Simple)
#### XML

```
<com.sdsmdg.harjot.rotatingtext.RotatingTextWrapper
        android:id="@+id/custom_switcher"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content" />
```

#### Java

```
RotatingTextWrapper rotatingTextWrapper = (RotatingTextWrapper) findViewById(R.id.custom_switcher);
rotatingTextWrapper.setSize(35);

Rotatable rotatable = new Rotatable(Color.parseColor("#FFA036"), 1000, "Word", "Word01", "Word02");
rotatable.setSize(35);
rotatable.setAnimationDuration(500);

rotatingTextWrapper.setContent("This is ?", rotatable);
```

#### Result
<img src="/screens/gif_example_1.gif"/>

## Example Usage 2 (Typeface + Interpolator)
#### XML

```
<com.sdsmdg.harjot.rotatingtext.RotatingTextWrapper
        android:id="@+id/custom_switcher"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content" />
```

#### Java

```
Typeface typeface = Typeface.createFromAsset(getAssets(), "fonts/Raleway-Light.ttf");
Typeface typeface2 = Typeface.createFromAsset(getAssets(), "fonts/Reckoner_Bold.ttf");

RotatingTextWrapper rotatingTextWrapper = (RotatingTextWrapper) findViewById(R.id.custom_switcher);
rotatingTextWrapper.setSize(35);
rotatingTextWrapper.setTypeface(typeface2);

Rotatable rotatable = new Rotatable(Color.parseColor("#FFA036"), 1000, "Word", "Word01", "Word02");
rotatable.setSize(35);
rotatable.setAnimationDuration(500);
rotatable.setTypeface(typeface);
rotatable.setInterpolator(new BounceInterpolator());

rotatingTextWrapper.setContent("This is ?", rotatable);
```

#### Result
<img src="/screens/gif_example_2.gif"/>

## Example Usage 3 (Multiple Rotatables)
#### XML

```
<com.sdsmdg.harjot.rotatingtext.RotatingTextWrapper
        android:id="@+id/custom_switcher"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content" />
```

#### Java

```
Typeface typeface = Typeface.createFromAsset(getAssets(), "fonts/Raleway-Light.ttf");
Typeface typeface2 = Typeface.createFromAsset(getAssets(), "fonts/Reckoner_Bold.ttf");

RotatingTextWrapper rotatingTextWrapper = (RotatingTextWrapper) findViewById(R.id.custom_switcher);
rotatingTextWrapper.setSize(35);
rotatingTextWrapper.setTypeface(typeface2);

Rotatable rotatable = new Rotatable(Color.parseColor("#FFA036"), 1000, "Word", "Word01", "Word02");
rotatable.setSize(35);
rotatable.setTypeface(typeface);
rotatable.setInterpolator(new AccelerateInterpolator());
rotatable.setAnimationDuration(500);

Rotatable rotatable2 = new Rotatable(Color.parseColor("#123456"), 1000, "Word03", "Word04", "Word05");
rotatable2.setSize(25);
rotatable2.setTypeface(typeface);
rotatable2.setInterpolator(new DecelerateInterpolator());
rotatable2.setAnimationDuration(500);

rotatingTextWrapper.setContent("This is ? and ?", rotatable, rotatable2);
```

#### Result
<img src="/screens/gif_example_3.gif"/>

# Documentation

Rotating text is made of two parts : `RotatingTextWrapper` and `Rotatable`. <br>
Each rotatable encapsulates the collection of words that are two be periodically switched and also defines various properties related to these words, like, size, color, animation interpolator etc.<br>
Each Rotatable must be a part of a `RotatingTextWrapper`. This defines the actual layout of the text and the positions of the rotating text.

For eg : `rotatingTextWrapper.setContent("This is ?", rotatble);`. Here the `?` denotes the postion of the `rotatable`.

## RotatingTextWrapper
|Property         |Function                |Description                             |
|-----------------|------------------------|----------------------------------------|
|Content               | setContent(...)                    | Set the actual content. Composed of a String and array of Rotatables. |
|Typeface              | setTypeface(...)                   | Set the typeface of the non-rotating text                     |
|Size                  | setSize(...)                       | Set the size of the non-rotating text                         |
|Pause                 | pause(x)                           | Method to pause the 'x'th rotatable                           |
|Resume                | resume(x)                          | Method to resume the 'x'th rotatable                          |

## Rotatable
|Property         |Function                |Description                             |
|-----------------|------------------------|----------------------------------------|
|Color                 | setColor(...)                      | Set the color of the rotating text associated with this rotatable     |
|Size                  | setSize(...)                       | Set the size of the rotating text associated with this rotatable      |
|Typeface              | setTypeface(...)                   | Set the typeface of the rotating text associated with this rotatable  |
|Interpolator          | setInterpolator(...)               | Set the animation interpolator used while switching text              |
|Update Duration       | setUpdateDuration(...)             | Set the interval between switching the words                          |
|Animation Duration    | setAnimationDuration(...)          | Set the duration of the switching animation                           |
|Center Align          | setCenter(...)                     |Align the rotating text to center of the textview if set to **true**   |


# License
RotatingText is licensed under `MIT license`. View [license](LICENSE.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Unreleased 

### Added
- Option for changing color of rotating texts by [@Anu-123-gif](https://github.com/Anu-123-gif) and [@dev-ritik](https://github.com/dev-ritik) in [9d8a330](https://github.com/mdg-iitr/RotatingText/commit/9d8a330fee558856c564db7318579038c8ca1c40)
- Stop rotation after one cycle by [@Anu-123-gif](https://github.com/Anu-123-gif) and [@dev-ritik](https://github.com/dev-ritik) in [8ed21ea](https://github.com/mdg-iitr/RotatingText/commit/8ed21ea7825f4f3a87af523e1a0693a78eaf646e)
- Method to add words to rotating texts added by [@dev-ritik](https://github.com/dev-ritik) in [118ab03](https://github.com/mdg-iitr/RotatingText/commit/118ab03d598cdb19e8ee2b8cff6586e7e23a19e4)
- onMeasure method added by [@dev-ritik](https://github.com/dev-ritik) in [dea14a2](https://github.com/mdg-iitr/RotatingText/commit/dea14a2099f27ba730b62be30b786c94cb56c01a)
- Adaptable function added by [@dev-ritik](https://github.com/dev-ritik) in [075b235](https://github.com/mdg-iitr/RotatingText/commit/075b23527408ba89e201fec3afea9364343573dc)
- Added requestLayout() calls by [@harjot-oberai](https://github.com/harjot-oberai) in [a0f8a1f](https://github.com/mdg-iitr/RotatingText/commit/a0f8a1fc763b25870edd8831468df25792f1f9b5)
- Added rx-android dependencies by [@harjot-oberai](https://github.com/harjot-oberai) in [c90af4c](https://github.com/mdg-iitr/RotatingText/commit/c90af4c25c0110586b34d55ff9c6466082a701a9)

### Changed
- Used interface to rename word by [@dev-ritik](https://github.com/dev-ritik) in [289c04f](https://github.com/mdg-iitr/RotatingText/commit/289c04f1ff91e6653c798c39629a9e17e73d7cf7)
- Renamed addWord method to replaceWord by [@harjot-oberai](https://github.com/harjot-oberai) in [36c31fd](https://github.com/mdg-iitr/RotatingText/commit/3646f79dfe79bfe21fe13cb81bc24a4eb0d2604d)
- "One line only" warning added in app by by [@dev-ritik](https://github.com/dev-ritik) in [075b235](https://github.com/mdg-iitr/RotatingText/commit/075b23527408ba89e201fec3afea9364343573dc)
- Updated dependencies by by [@harjot-oberai](https://github.com/harjot-oberai) in [a46ceb9](https://github.com/mdg-iitr/RotatingText/commit/a46ceb945d6f8ba5f8d6dc6e1eccc45fd556780c)

### Fixed
- Code improved and one word bug fixed by [@dev-ritik](https://github.com/dev-ritik) in [f29c8b7](https://github.com/mdg-iitr/RotatingText/commit/f29c8b7bc9ce1caf79992af12e6690a4b5932509)


app/src/main/java/com/sdsmdg/harjot/rotatingtextlibrary/MainActivity.java
package com.sdsmdg.harjot.rotatingtextlibrary;

import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;

import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;

import com.sdsmdg.harjot.rotatingtext.RotatingTextSwitcher;
import com.sdsmdg.harjot.rotatingtext.RotatingTextWrapper;
import com.sdsmdg.harjot.rotatingtext.models.Rotatable;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    RotatingTextWrapper rotatingTextWrapper;
    RotatingTextSwitcher rotatingTextSwitcher;
    Rotatable rotatable, rotatable2;
    Spinner s1;
    String word;
    int nCycles;
    EditText enterCycles;
    EditText e1;

    Button button;
    Integer[] color = new Integer[]{Color.BLUE,Color.MAGENTA,Color.RED};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rotatingTextSwitcher = new RotatingTextSwitcher(this);
        Typeface typeface = Typeface.createFromAsset(getAssets(), "fonts/Raleway-Light.ttf");
        Typeface typeface2 = Typeface.createFromAsset(getAssets(), "fonts/Reckoner_Bold.ttf");

        rotatingTextWrapper = findViewById(R.id.custom_switcher);
        rotatingTextWrapper.setSize(30);
        rotatingTextWrapper.setTypeface(typeface2);

//        rotatable = new Rotatable(Color.parseColor("#FFA036"), 1000, "Word00", "Word01", "Word02");
        rotatable = new Rotatable(color , 1000, "rotating", "text", "library");
        rotatable.setSize(25);
        rotatable.setTypeface(typeface);
        rotatable.setInterpolator(new AccelerateInterpolator());
        rotatable.setAnimationDuration(500);

        rotatable2 = new Rotatable(Color.parseColor("#123456"), 1000, "word", "word01", "word02");
        rotatable2.setSize(25);
        rotatable2.setTypeface(typeface);
        rotatable2.setInterpolator(new DecelerateInterpolator());
        rotatable2.setAnimationDuration(500);


        word = rotatable.getTextAt(0);

        rotatingTextWrapper.setContent("?abc ? abc", rotatable, rotatable2);
//        rotatingTextWrapper.setContent("? abc", rotatable);

        s1 = (Spinner) findViewById(R.id.spinner);



        List<Integer> list = new ArrayList<Integer>();
        list.add(1);
        list.add(2);
        list.add(3);

        ArrayAdapter<Integer> dataAdapter = new ArrayAdapter<Integer>(this,
                android.R.layout.simple_spinner_item, list);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        s1.setAdapter(dataAdapter);
        s1.setSelection(0);

        e1 = (EditText) findViewById(R.id.replaceWord);

        button = findViewById(R.id.pause_button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!rotatingTextWrapper.getSwitcherList().get(0).isPaused()) {
                    rotatingTextWrapper.pause(0);
                } else {
                    rotatingTextWrapper.resume(0);
                }
            }
        });

        Button button2 = findViewById(R.id.set_button);
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                enterCycles = (EditText) findViewById(R.id.enterCycles);


                if(!enterCycles.getText().toString().equals("") && enterCycles.getText() != null) {
                    nCycles = Integer.parseInt(enterCycles.getText().toString());
                }

                rotatable.setInitialWord(word);
                rotatable.setCycles(nCycles);

                if (rotatingTextWrapper.getSwitcherList().get(0).isPaused()) {
                    rotatingTextWrapper.resume(0);
                }
            }
        });

        Button buttonChange = findViewById(R.id.change);
        buttonChange.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean apply = rotatable.getApplyHorizontal();
                rotatable.setApplyHorizontal(!apply);
                rotatable2.setApplyHorizontal(!apply);
            }
        });
    }

    public void replaceWord(View view) {
        String newWord = e1.getText().toString();
        if (TextUtils.isEmpty(newWord)) e1.setText("can't be left empty");
        else if (newWord.contains("\n")) e1.setText("one line only");
        else {
            rotatingTextWrapper.setAdaptable(true);
            rotatingTextWrapper.addWord(0, (int) s1.getSelectedItem() - 1, newWord);
        }

    }
}

